polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp

```diff
@@ -599,19 +592,14 @@ bool ArgumentPromotionPass::isCandidateCallable(
   // in a function that is called directly by a GPU kernel.
   // TODO: Could generalize by checking that the call chain from the GPU kernel
   // are all candidates.
-  for (Operation *call : userMap.getUsers(callableOp)) {
-    if (call->getParentOfType<gpu::GPUFuncOp>())
-      continue;
-
-    auto funcOp = call->getParentOfType<FunctionOpInterface>();
-    if (!funcOp || !llvm::all_of(userMap.getUsers(funcOp), [](Operation *call) {
-          return call->getParentOfType<gpu::GPUFuncOp>();
-        })) {
-      LLVM_DEBUG(llvm::dbgs().indent(2)
-                 << "not a candidate: found call site that is neither in a GPU "
-                    "kernel nor in a function called by a GPU kernel\n");
-      return false;
-    }
+  int maxDepth = getMaxDepthFromGPUKernel(funcOp);
+  assert(maxDepth > 0 && "Expecting func to be called from a GPU kernel and "
+                         "not itself a GPU kernel");
+  if (maxDepth > 2) {
```

Comment:
I think this should be  `maxDepth >= 2`  because a function called directly by a gpu kernel has depth = 0, and a function called by a function that is itself called by a gpu kernel has depth = 1.

I think we have an issue with this scheme though because is possible for a function `foo` to be called in more than one calling context. Assume K1, K2, K3 are gpu kernels and we have the following calling contexts: 

    K1 -> foo     (here foo's depth = 0)
    K2 -> bar -> foo   (here bar's depth = 0, foo's depth = 1)
    K3 -> faa -> bar -> foo (here bar's depth = 1, foo's depth = 2) 

So, in general, a function might have different depths based on context.  We should associate the entry point of the call chain (e.g. the gpu kernel) to the depth of a function.

Reply:
With the current implementation, gpu kernel itself is depth = 0, called directly by a gpu kernel has depth = 1, so function called by another function that directly called by a gpu kernel has depth = 2.
That's why we ask for the maximum depth, in your example, should be 3, so foo is not a candidate.

Reply:
Ok starting from 1 for a function called directly from a GPU kernel is OK. However what about differentiating the calling context ? Stating that a function has depth 1 without without specifying w.r.t which kernel is not sufficient.

Reply:
After offline discussion this function is supposed to return the max depth from *any* GPU kernel in a gpu module. Therefore it need not track the calling context as I first assumed. Please change the name of the function `getMaxDepthFromGPUKernel` to `getMaxDepthFromAnyGPUKernel`